PYTHON-2252 Add examples and documentation for new UUID behavior

prashantmital opened the pull request.

behackett commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,290 @@
+   written to MongoDB by existing applications that use the Python driver
+   and don't explicitly set a UUID representation.
+
+.. attention:: As of PyMongo 3.11.0,
```
This implies to me that this is new in 3.11.0. Better to say that PYTHON_LEGACY is the default and has been since PyMongo 2.9, the version it was introduced in.

behackett commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,290 @@
+
+.. _configuring-uuid-representation:
+
+Configuring a UUID Representation
```
Somewhere in here you should talk about round tripping data, and you might want to warn about it. It's easy for an application to read and update a record accidentally changing the byte order of the object in MongoDB.

ShaneHarvey replied:
@prashantmital, an example of changing byte order would be round tripping a Binary subtype 4 with uuid representation C# or Java. The subtype 4 would be decoded to a native UUID and then encoded as Binary subtype 3 with **different** bytes.

prashantmital replied:
As clarified on Slack, the bytes are not different, but the subtype is. This is reflected in the new example.

ShaneHarvey replied:
No, the bytes would also change:
```
>>> u = uuid.UUID('00112233445566778899AABBCCDDEEFF')
>>> b_standard = bson.encode({'u': u}, codec_options=CodecOptions(uuid_representation=UuidRepresentation.STANDARD))
>>> b_standard
b'\x1d\x00\x00\x00\x05u\x00\x10\x00\x00\x00\x04\x00\x11"3DUfw\x88\x99\xaa\xbb\xcc\xdd\xee\xff\x00'
>>> decoded = bson.decode(b_standard, codec_options=CodecOptions(uuid_representation=UuidRepresentation.JAVA_LEGACY))
>>> decoded
{'u': UUID('00112233-4455-6677-8899-aabbccddeeff')}
>>> u
UUID('00112233-4455-6677-8899-aabbccddeeff')
>>> # The UUID's subtype AND data change here:
>>> b_legacy = bson.encode(decoded, codec_options=CodecOptions(uuid_representation=UuidRepresentation.JAVA_LEGACY))
>>> b_legacy
b'\x1d\x00\x00\x00\x05u\x00\x10\x00\x00\x00\x03wfUD3"\x11\x00\xff\xee\xdd\xcc\xbb\xaa\x99\x88\x00'
>>> b_standard
b'\x1d\x00\x00\x00\x05u\x00\x10\x00\x00\x00\x04\x00\x11"3DUfw\x88\x99\xaa\xbb\xcc\xdd\xee\xff\x00'
>>> decoded = bson.decode(b_legacy, codec_options=CodecOptions(uuid_representation=UuidRepresentation.JAVA_LEGACY))
>>> decoded
{'u': UUID('00112233-4455-6677-8899-aabbccddeeff')}
>>> decoded = bson.decode(b_legacy, codec_options=CodecOptions(uuid_representation=UuidRepresentation.STANDARD))
>>> decoded
{'u': UUID('77665544-3322-1100-ffee-ddccbbaa9988')}
```

prashantmital replied:
I am don't follow this completely. Will need to discuss.

ShaneHarvey replied:
Here's the same example written from a different perspective. App `A` inserts a document with STANDARD. App `B` find and replaces the same document with JAVA_LEGACY.
```python
>>> coll_standard = client.t.get_collection('t', codec_options=CodecOptions(uuid_representation=UuidRepresentation.STANDARD))
>>> coll_java = client.t.get_collection('t', codec_options=CodecOptions(uuid_representation=UuidRepresentation.JAVA_LEGACY))
>>> coll_raw = client.t.get_collection('t', codec_options=CodecOptions(document_class=RawBSONDocument))
>>> u = UUID('00112233-4455-6677-8899-aabbccddeeff')
>>> coll_standard.insert_one({'_id': 1, 'u': u})
<pymongo.results.InsertOneResult object at 0x7fe15e453500>
>>> # Raw bytes on the server:
>>> coll_raw.find_one({'_id':1})
RawBSONDocument(b'&\x00\x00\x00\x10_id\x00\x01\x00\x00\x00\x05u\x00\x10\x00\x00\x00\x04\x00\x11"3DUfw\x88\x99\xaa\xbb\xcc\xdd\xee\xff\x00', ...)
>>> # Find/replace the document from a legacy Java app:
>>> doc = coll_java.find_one({'_id': 1})
>>> coll_java.replace_one({'_id': 1}, doc)
<pymongo.results.UpdateResult object at 0x7fe15e453b00>
>>> # Raw bytes on the server have changed:
>>> coll_raw.find_one({'_id':1})
RawBSONDocument(b'&\x00\x00\x00\x10_id\x00\x01\x00\x00\x00\x05u\x00\x10\x00\x00\x00\x03wfUD3"\x11\x00\xff\xee\xdd\xcc\xbb\xaa\x99\x88\x00', ...)
```

prashantmital replied:
Added an example for this.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,290 @@
+:class:`~bson.binary.Binary` subtype 4 objects.
+
+
+The ``UNSPECIFIED`` UUID Representation
```
Can you add a code example of what happens when you read a document containing a UUID with UNSPECIFIED and another example of what happens when you attempt (and fail) to write a document with a UUID? I think it would be nice to include the full error traceback here.

behackett replied:
You should also mention, if you haven't already - I might have missed it, that UNSPECIFIED will become the default in PyMongo 4.0.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,290 @@
+
+    from uuid import UUID
+    collection = client.example_db.uuid_test
+    results = collection.find({'_id': UUID('00112233-4455-6677-8899-aabbccddeeff')})
```
suggest find_one since this is a find-by-id with at most one result.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,290 @@
+  BSON Binary. The following query, on the other hand, will successfully find
+  this document::
+
+    results = collection.find({'_id': UUID('33221100-5544-7766-8899-aabbccddeeff')})
```
suggest find_one since this is a find-by-id with at most one result.

prashantmital commented:
I addressed all comments other than 

> Somewhere in here you should talk about round tripping data, and you might want to warn about it. It's easy for an application to read and update a record accidentally changing the byte order of the object in MongoDB.

I was unable to come up with an example for this.

prashantmital commented:
Ready for final review.

behackett commented on doc/examples/uuid.rst:
```diff
@@ -281,13 +322,43 @@ when encoding UUIDs from all drivers. UUIDs written by a driver with this
+.. attention:: In PyMongo 3.x, applications can end up inadvertently changing
+   the :class:`~bson.binary.Binary` subtype of a UUID field when round-tripping
+   documents with legacy-encoded UUIDs. This happens because in PyMongo 3.x,
+   the ``STANDARD`` representation decodes all UUIDs (subtypes 3 and 4) as
```
The opposite is also true, I think? If you retrieve a document with subtype 4 fields but have any of the legacy representations configured?

prashantmital replied:
Ah, that's correct. I will break this out into a separate section dedicated to problems that can be caused by using the incorrect representation.

prashantmital replied:
Done.

prashantmital commented:
Ready for another look.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,472 @@
+a UUID When using this representation returns a :class:`~bson.binary.Binary`
+object instead. If required, users can coerce the decoded
+:class:`~bson.binary.Binary` objects into native UUIDs using the
+:meth:`~bson.binary.Binary.as_uuid` method and specifyin the appropriate
```
specifyin -> specifying

prashantmital replied:
Done

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,472 @@
+The :data:`~bson.binary.UuidRepresentation.UNSPECIFIED` representation
+prevents the incorrect interpretation of UUID bytes by stopping short of
+automatically converting UUID fields in BSON to native UUID types. Loading
+a UUID When using this representation returns a :class:`~bson.binary.Binary`
```
When -> when

prashantmital replied:
Done.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,472 @@
+  # Using UuidRepresentation.CSHARP_LEGACY
+  csharp_opts = CodecOptions(uuid_representation=UuidRepresentation.CSHARP_LEGACY)
+
+  # Store a C#-formatted UUID
```
"Store a legacy C#-formatted UUID"

prashantmital replied:
Done.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,472 @@
+  unspec_collection = client.testdb.get_collection('test', codec_options=unspec_opts)
+
+  # UUID fields are decoded as Binary when UuidRepresentation.UNSPECIFIED is configured
+  uuid_1 = uuid4()
```
Can we remove this line?

prashantmital replied:
Yep. Cruft from copy-pasting. Removed.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,472 @@
+  unspec_collection.insert_one({'_id': 'bar', 'uuid': uuid4()})
+  Traceback (most recent call last):
+  ...
+  ValueError: cannot encode native uuid.UUID with UuidRepresentation.UNSPECIFIED. UUIDs can be manually converted to bson.Binary instances using bson.Binary.from_uuid() or a different UuidRepresentation can be configured.
```
Should this error message link to this documentation page? I'm wondering if we can help point the user to the right place.

prashantmital replied:
Not a huge fan of hardcoding links into the codebase, but we can do it if you feel strongly about it.

ShaneHarvey replied:
What about something like "see the documentation for UuidRepresentation for more information".

prashantmital replied:
Sounds good... Making this change.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,472 @@
+^^^^^^^^^^^^^^^^^
+
+.. attention:: This uuid representation should be used when reading UUIDs
+   written to MongoDB by the C# driver without an explicitly configured UUID
```
This wording makes it seem like all C# apps have this problem. Should we clarify that it's only legacy C# apps (which don't use STANDARD)?

prashantmital replied:
Done.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,472 @@
+^^^^^^^^^^^^^^^
+
+.. attention:: This uuid representation should be used when reading UUIDs
+   written to MongoDB by the Java driver without an explicitly configured UUID
```
This wording makes it seem like all Java apps have this problem. Should we clarify that it's only legacy Java apps (which don't use STANDARD)?

prashantmital replied:
Done.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,472 @@
+==================
+
+PyMongo ships with built-in support for dealing with UUID types.
+It is trivially simple to store native :class:`uuid.UUID` objects
```
I feel like saying "trivially simple" before a wall of text dedicated to the complexity of UUIDs is odd.

prashantmital replied:
Changed to 'straightforward'

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,472 @@
+:class:`~bson.binary.UuidRepresentation` configuration option to maintain
+cross-language compatibility.
+
+.. attention:: Applications that do not share a MongoDB deployment with
```
Applications -> New applications?

prashantmital replied:
Done.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,472 @@
+
+Consider the following situation:
+
+* Application ``M`` written in C# generates a UUID and uses it as the ``_id``
```
Can you change `M` and `N` to `A` and `B` (or `C` for C# and `P` for Python)? `M` and `N` sound and look very similar.

prashantmital replied:
Done. Used C and P

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,472 @@
+that replacing ``PYTHON_LEGACY`` by ``JAVA_LEGACY`` or ``CSHARP_LEGACY`` in
+the above example produces the same behavior.**
+
+.. note:: This will not be an issue in PyMongo>=4 as starting in that version,
```
"Starting in PyMongo 4.0 this issue will be resolved"

prashantmital replied:
done.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -0,0 +1,472 @@
+   :class:`~bson.binary.Binary` objects of subtype 3 (instead of
+   :class:`uuid.UUID`), and each of the ``LEGACY_*`` representations will
+   decode Binary subtype 4 fields to :class:`~bson.binary.Binary` objects of
+   subtype 4 (instead of :class:`uuid.UUID`). This will prevent
```
What will this prevent?

prashantmital replied:
Nothing to add here. Removed the sentence stub.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -164,11 +165,48 @@ In this example, round-tripping the document using the incorrect
+
+  # Round-tripping the retrieved document silently changes the Binary bytes and subtype
+  java_collection.replace_one({'_id': 'baz'}, doc)
+  assert collection.find_one({'uuid': Binary(input_uuid.bytes, 3)}) is None
```
I think this should be `Binary(input_uuid.bytes, 4)`

prashantmital replied:
Done - added the additional assert.

ShaneHarvey commented on doc/examples/uuid.rst:
```diff
@@ -164,11 +165,48 @@ In this example, round-tripping the document using the incorrect
+used to round-trip UUIDs written with ``STANDARD``. When the situation is
+reversed - i.e. when the original document is written using ``STANDARD``
+and then round-tripped using ``CSHARP_LEGACY`` or ``JAVA_LEGACY`` -
+only the :class:`~bson.binary.Binary` subtype is changed.**
```
The reverse situation would also change the bytes stored in the database. Also I think the "i.e. when the original document is written using ``STANDARD``..." part is written backwards.

prashantmital replied:
You are right about it being backwards. 
However, I think that when the original UUID is stored with *_LEGACY and then round-tripped using STANDARD only the subtype would change - STANDARD reads subtype 3 as PYTHON_LEGACY and writes them back as subtype 4 without changing byte-order.

ShaneHarvey replied:
Ah right. So the bytes on the server don't change, but the python UUID instance will be different.
